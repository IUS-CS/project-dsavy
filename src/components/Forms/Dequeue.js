import React from 'react'
import { Button} from 'react-bootstrap';

const Dequeue = ({dequeue, title}) => {
  
  const onClick = (e) => {
      dequeue()
  }

  return (
    <Button className="my-1" onClick={onClick}>{title}</Button>
  )

}

export default Dequeue