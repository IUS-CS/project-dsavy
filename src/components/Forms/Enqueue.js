import React, {useState} from 'react'
import {Button} from 'react-bootstrap';

const Enqueue = ({enqueue, title}) => {
  const [entry, setEntry] = useState('')
  
  const onSubmit = (e) => {
      e.preventDefault()
      enqueue(entry)
      setEntry('')
  }

  return (
    <form onSubmit={onSubmit}>
      <Button type="submit" className="my-1">{title}</Button>
      <input title="queue-input" className='enqueue' type='number' value={entry} onChange={(e) => setEntry(e.target.value)} />
    </form>
  )
}

export default Enqueue