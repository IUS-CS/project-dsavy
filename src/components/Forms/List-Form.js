import React, { useState } from 'react';
import { Form, Row, Col, Button, Container } from 'react-bootstrap';
import Check from '../Char-Limit-Error';

const ListForm = ({list, onListChange}) => {
    const handleChange = e => {
        onListChange(e);
    }

    const [validLength, setValidLength] = useState(true);

    const onInsert = e => {
        e.preventDefault()
        const formData = new FormData(e.target),
              formDataObj = Object.fromEntries(formData.entries());

        // Check length of new element is not over limit
        if (formDataObj.Value.length > 6) {
            setValidLength(false);
        }
        else {
            setValidLength(true);
            if (formDataObj.Place === 'Tail'){
                handleChange(list => list.concat({key: list.length, item: formDataObj.Value}));
            }
            else if (formDataObj.Place === '0'){
                let newList = [{key: 0, item: formDataObj.Value}];
                for (let i = 0; i < list.length; i++) {
                    list[i].key = i + 1;
                    newList.push(list[i]);
                }
                handleChange(list => newList);
            }
            else {
                let currentKey = parseInt(formDataObj.Place);
                let newItem = {key: currentKey, item: formDataObj.Value};
                let newList = [];
                
                for (let i = 0; i < list.length + 1; i++) {
                    if (currentKey === i) {
                        newList.push(newItem);
                    }
                    else if (currentKey > i) {
                        list[i].key = i;
                        newList.push(list[i]);
                    }
                    else if (currentKey < i) {
                        list[i - 1].key = i;
                        newList.push(list[i - 1]);
                    }
                }
                handleChange(list => newList);
            }
            console.log(list);
          }
        }
        

    const onRemove = e => {
        e.preventDefault()
        const formData = new FormData(e.target),
              formDataObj = Object.fromEntries(formData.entries());
        console.log(formDataObj);
        let listCopy = [...list];
        if (formDataObj.Place === 'Tail') {
            listCopy.splice((listCopy.length - 1), 1);
        }
        else {
            listCopy.splice(formDataObj.Place, 1);
        }
        
        handleChange(listCopy);
        console.log(list);
    }


    return (
        <div>
        <Check input={validLength}></Check>
        <Container fluid style={{paddingTop:60}}>
            <Row style={{justifyContent:'center', fontSize:20, fontWeight:500, paddingTop:15}}>
                Insert
            </Row>
            <Row style={{justifyContent:'center'}}>
            <Form inline onSubmit={onInsert}>
                <Form.Control className="my-1 mr-2" type="text" name="Value" placeholder="Value" />
                <Form.Label className="my-1 mr-2" htmlFor="inlineFormCustomSelectPref">
                    At Position
                </Form.Label>
                <Form.Control
                    as="select"
                    className="my-1 mr-sm-2"
                    id="inlineFormCustomSelectPref"
                    name="Place"
                >
                    <option value="Tail">Tail</option>
                    <option value="0">Head</option>
                    <option value="1">One</option>
                    <option value="2">Two</option>
                    <option value="3">Three</option>
                    
                </Form.Control>
                <Button type="submit" className="my-1">Insert</Button>
            </Form>
            </Row>
            <Row style={{justifyContent:'center', fontSize:20, fontWeight:500, paddingTop:15}}>
                Remove
            </Row>
            <Row style={{justifyContent:'center'}}>
                <Form inline onSubmit={onRemove}>
                    <Form.Label className="my-1 mr-2" htmlFor="inlineFormCustomSelectPref">
                        Element at Position
                    </Form.Label>
                    <Form.Control
                        as="select"
                        className="my-1 mr-sm-2"
                        id="inlineFormCustomSelectPref"
                        name="Place"
                    >
                        <option value="0">Head</option>
                        <option value="1">One</option>
                        <option value="2">Two</option>
                        <option value="3">Three</option>
                        <option value="Tail">Tail</option>

                    </Form.Control>
                    <Button type="submit" className="my-1" style={{}}>Remove</Button>
                </Form>
            </Row>
        </Container>
        </div>
    )
}

export default ListForm;