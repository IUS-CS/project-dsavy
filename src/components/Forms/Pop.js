import React from 'react'
import { Button} from 'react-bootstrap';

const Pop = ({pop}) => {
  
  const onClick = (e) => {
      e.preventDefault()
      pop()
  }

  return (
    <Button className="my-1" onClick={onClick}>Pop</Button>
  )

}

export default Pop