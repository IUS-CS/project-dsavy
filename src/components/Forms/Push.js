import React, {useState} from 'react'
import { Button} from 'react-bootstrap';

const Push = ({push}) => {
  const [entry, setEntry] = useState('')
  
  const onSubmit = (e) => {
      e.preventDefault()
      push({entry})
      setEntry('')
  }

  return (
    <form onSubmit={onSubmit}>
      <Button type="submit" className="my-1">Push</Button>
      <input className='push' type='number' value={entry} onChange={(e) => setEntry(e.target.value)} />
    </form>
  )
}

export default Push