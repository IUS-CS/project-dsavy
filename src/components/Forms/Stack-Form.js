import React, {useState } from 'react';
import { Form, Row, Col, Button, Container } from 'react-bootstrap';
import {useTransition, animated} from 'react-spring';

const StackNode = (props) => {
    const styleSheet = {
        position: 'absolute',
        background: "linear-gradient(#58C2ED, #1B85DC)",
        width: '18rem',
        height: '3rem',
        border: 'solid',
        bottom: `${props.id*3}rem`,
        right: '-2.5px',
        borderTop: '0'
    }
    return (
        <div id={`${props.id}`} style={styleSheet}>
            <h4 style={{textAlign: 'center', color: 'white', fontWeight: '550'}}>{props.value}</h4>
        </div>
    );
}

const ListForm = () => {
    const [stack, setStack] = useState([]);
    const [value, setValue] = useState('');
    const [count, setCount] = useState(0);

    function onPush(event) {
        event.preventDefault();
        setStack(push(stack, value));
    }

    function push(array, value) {
        if (value == '') {
            alert('Enter number');
            return array;
        }
      
        if (array.length > 9) {
            alert('Stack is at max length');
            return array;
        }
      
        if (Math.abs(value) > 999) {
            alert('Number must be 3 digits or less');
            return array;
        }
          
        setCount(count + 1);
        return array.concat([<StackNode id={count} value={value} key={count}/>]);
    }
    
    function handleChange(event) {
        setValue(event.target.value);
    }
    
    function onPop(event) {
        event.preventDefault();
        if (stack.length !== 0) {
            setCount(count - 1);
            setStack(pop(stack));
        }
    }

    function onPeek(event) {
        event.preventDefault();
        // change the color
        document.getElementById(`${stack.length-1}`).style.background = "linear-gradient(#ff9d00, #ffb61785)";
        setTimeout(() => {
            // reset the color after a second
            document.getElementById(`${stack.length-1}`).style.background = "linear-gradient(#58C2ED, #1B85DC)";
        }, 1000);
    }

    function pop(array) {
        if (array.length == 0) {return array}
        array.pop()
        return array
    }

    const transition = useTransition(stack, item => item.key, {
        from: {opacity: 0.5, transform: 'translate3d(0,10rem,0)'},
        enter: {opacity: 1, transition: 'opacity .25s', transform: 'translate3d(0,28rem,0)'},
        leave: {opacity: 0, transition: 'opacity 1s', color: 'red', transform: 'translate3d(0,-28rem,0)'},
        config: {
            duration: 500,
        }
    })

    const showStack = () => {
        return (
            transition.map(({item, props, key}) => 
            <animated.div key={key} style={props}>
                {item}
            </animated.div>)
        )
    }
    
    return (
        <Container fluid style={{paddingTop:60, paddingLeft: 30}}>
            <Row style={{justifyContent:'left', fontSize:20, fontWeight:500, paddingTop:15}}>
                Insert
            </Row>
            <Row style={{justifyContent:'left'}}>
            <Form data-testid="values">
                <Form.Control className="my-1 mr-2" type="text" name="value" placeholder="Value" onChange={handleChange} value={value}/>
                <Button onClick={onPush} type="submit" className="my-1">Push</Button>
            </Form>
            </Row>

            <Row style={{justifyContent:'left', fontSize:20, fontWeight:500, paddingTop:15}}>
                Remove
            </Row>
            <Row style={{justifyContent:'left'}}>
                <Form inline onSubmit={onPop}>
                    <Button type="submit" className="my-1" style={{}}>Pop</Button>
                </Form>
            </Row>

            <Row style={{justifyContent:'left', fontSize:20, fontWeight:500, paddingTop:15}}>
                Look
            </Row>
            <Row style={{justifyContent:'left'}}>
                <Form inline onSubmit={onPeek}>
                    <Button type="submit" className="my-1" style={{}}>Peek</Button>
                </Form>
            </Row>
            
            {/* this div holds the stack nodes */}
            <div id="stackbox" style={
                    {width: '18rem', height: '28rem', position: 'relative', right: '9rem',
                    bottom: '12rem', left: '42rem', border: 'solid', borderTop: 0}}>
                {showStack()}
            </div>
        </Container>
    )
}

export default ListForm;