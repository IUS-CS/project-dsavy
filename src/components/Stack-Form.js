import React, { useState } from 'react';
import { Form, Row, Col, Button, Container } from 'react-bootstrap';

const ListForm = ({list, onListChange}) => {
    const [queue, setQueue] = useState([])

    const onPush = ({entry}) => {
      setQueue(queue.concat(entry))
    }
  
    const onPop = () => {
      setQueue(queue.slice(1))
    }

    return (
        <Container fluid style={{paddingTop:60, paddingLeft: 60}}>
            <Row style={{justifyContent:'left', fontSize:20, fontWeight:500, paddingTop:15}}>
                Insert
            </Row>
            <Row style={{justifyContent:'left'}}>
            <Form inline onSubmit={onPush}>
                <Form.Control className="my-1 mr-2" type="text" name="Value" placeholder="Value" />
                
                <Button type="submit" className="my-1">Push</Button>
            </Form>
            </Row>
            <Row style={{justifyContent:'left', fontSize:20, fontWeight:500, paddingTop:15}}>
                Remove
            </Row>
            <Row style={{justifyContent:'left'}}>
                <Form inline onSubmit={onPop}>
                    <Button type="submit" className="my-1" style={{marginRight: 5}}>Pop</Button>
                    <Form.Label className="my-1 mr-2" htmlFor="inlineFormCustomSelectPref">
                        Element on the Top
                    </Form.Label>
                </Form>
            </Row>
            <Row style={{justifyContent:'left', fontSize:20, fontWeight:500, paddingTop:15}}>
                Peek
            </Row>
            <Row style={{justifyContent:'left'}}>
                <Form inline onSubmit={onRemove}>
                    <Button type="submit" className="my-1" style={{marginRight: 5}}>Pop</Button>
                    <Form.Label className="my-1 mr-2" htmlFor="inlineFormCustomSelectPref">
                        Element on the Top
                    </Form.Label>
                </Form>
            </Row>
        </Container>
    )
}

export default ListForm;