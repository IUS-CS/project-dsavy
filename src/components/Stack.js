import React, {useState} from 'react';
import { useSpring, animated } from 'react-spring';
import StackForm from './Forms/Stack-Form';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';

const Stack = () => {
    return (
        <animated.div>
            <StackForm />

        </animated.div>
    );
}

export default Stack