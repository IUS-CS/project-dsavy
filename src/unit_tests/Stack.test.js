import React from 'react';
import Stack from 'components/Stack';
import 'components/Forms/Stack-Form'
import renderer from 'react-test-renderer';
import { Form, Row, Col, Button, Container } from 'react-bootstrap';
import { fireEvent, getByDisplayValue, getByLabelText, getByPlaceholderText, getByRole, getByText, render, screen } from '@testing-library/react';

test("Stack page renders correctly", () => {
    const tree = renderer.create(<Stack/>).toJSON()
    expect(tree).toMatchSnapshot()
});

describe('Testing push', () => {
    const push = (stack, number) => {
        if (number == '') {
          return stack
        }
        
        if (stack.length > 9) {
          return stack
        }
    
        if (Math.abs(number) > 999) {
          return stack
        }
        
        stack.push(number)
        return stack
    }

    it('push correctly', ()=>{
      let array = [1]
      expect(push(array,4)).toStrictEqual([1,4]);
    })

    it('push does not add, number too large', ()=>{
      let array = [1]
      expect(push(array,1000)).toStrictEqual([1]);
    })
    
    it('adds to empty stack', ()=>{
      let array = []
      expect(push(array,950)).toStrictEqual([950]);
    })

    it('Fails to add, stack at max length', ()=>{
      let array = [1,2,3,4,5,6,7,8,9,10]
      expect(push(array,950)).toStrictEqual([1,2,3,4,5,6,7,8,9,10]);
    })

    it('Fails to add, no entry', ()=>{
        let array = [1,2,3,4,5,6,7,8]
        expect(push(array,'')).toStrictEqual([1,2,3,4,5,6,7,8]);
      })
});

describe('Testing pop', () => {
    function pop(array) {
        if (array.length == 0) {return array}
        array.pop()
        return array
    }

    it('pops correctly', ()=>{
      let array = [1,3,4]
      expect(pop(array,4)).toStrictEqual([1,3]);
    })

});

// UI Testing
test("allows users to insert items in stack", () => {
  const wrapper = render(<Stack/>);

  const input = screen.getByPlaceholderText('Value')
  fireEvent.change(input, { target: { value: "one" } })
  fireEvent.click(screen.getByRole('button', {name: "Push"}))

  screen.getByText('one')
})