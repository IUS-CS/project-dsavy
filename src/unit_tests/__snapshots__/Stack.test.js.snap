// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack page renders correctly 1`] = `
<div>
  <div
    className="container-fluid"
    style={
      Object {
        "paddingLeft": 30,
        "paddingTop": 60,
      }
    }
  >
    <div
      className="row"
      style={
        Object {
          "fontSize": 20,
          "fontWeight": 500,
          "justifyContent": "left",
          "paddingTop": 15,
        }
      }
    >
      Insert
    </div>
    <div
      className="row"
      style={
        Object {
          "justifyContent": "left",
        }
      }
    >
      <form
        className=""
      >
        <input
          className="my-1 mr-2 form-control"
          name="value"
          onChange={[Function]}
          placeholder="Value"
          type="text"
          value=""
        />
        <button
          className="my-1 btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="submit"
        >
          Push
        </button>
      </form>
    </div>
    <div
      className="row"
      style={
        Object {
          "fontSize": 20,
          "fontWeight": 500,
          "justifyContent": "left",
          "paddingTop": 15,
        }
      }
    >
      Remove
    </div>
    <div
      className="row"
      style={
        Object {
          "justifyContent": "left",
        }
      }
    >
      <form
        className="form-inline"
        onSubmit={[Function]}
      >
        <button
          className="my-1 btn btn-primary"
          disabled={false}
          style={Object {}}
          type="submit"
        >
          Pop
        </button>
        <label
          className="my-1 mr-2 form-label"
        >
          The Element on Top
        </label>
      </form>
    </div>
    <div
      id="stackbox"
      style={
        Object {
          "border": "solid",
          "borderTop": 0,
          "bottom": "12rem",
          "height": "28rem",
          "left": "42rem",
          "marginBottom": "2rem",
          "position": "relative",
          "right": "9rem",
          "width": "18rem",
        }
      }
    />
  </div>
</div>
`;
